drivers/docker/driver.go

```diff
@@ -372,29 +387,34 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
 
 	startAttempts := 0
 CREATE:
-	container, err := d.createContainer(dockerClient, containerCfg, driverConfig.Image)
+	createdContainer, err := d.createContainer(dockerClient, containerCfg, driverConfig.Image)
 	if err != nil {
 		d.logger.Error("failed to create container", "error", err)
-		dockerClient.RemoveContainer(docker.RemoveContainerOptions{
-			ID:    containerCfg.Name,
-			Force: true,
-		})
+		if createdContainer != nil {
+			err := dockerClient.ContainerRemove(d.ctx, createdContainer.ID, containerapi.RemoveOptions{Force: true})
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to remove container %s: %v", createdContainer.ID, err)
+			}
+		}
 		return nil, nil, nstructs.WrapRecoverable(fmt.Sprintf("failed to create container: %v", err), err)
 	}
 
-	d.logger.Info("created container", "container_id", container.ID)
+	d.logger.Info("created container", "container_id", createdContainer.ID)
 
 	// We don't need to start the container if the container is already running
 	// since we don't create containers which are already present on the host
 	// and are running
+	container, err := dockerClient.ContainerInspect(d.ctx, createdContainer.ID)
+	if err != nil {
+		d.logger.Error("failed to inspect created container", "error", err)
+		return nil, nil, nstructs.WrapRecoverable(fmt.Sprintf("failed to create container: %v", err), err)
+	}
```

Comment:
Don't we already get a container with a valid `ID` and `State` fields (I think those are the only two we query?) from the `createContainer` method? What does re-querying the API get us? And if it _is_ necessary, should we push it down into the `createContainer` method?

Reply:
oof, it's a snafu that happened because initial refactoring had the `createContainer` return the wrong type. I fixed in [55f9670](https://github.com/hashicorp/nomad/pull/23966/commits/55f96706baed8ea17e27f9843a0f6c69ef6bd9d4)